### PR TITLE
Add SliverList

### PR DESCRIPTION
**Pull request by abarth**
A SliverList is a linear layout of box children in a viewport that all
have a common, fixed extent along the scroll axis. The layout is similar
to a SliverBlock but more efficient.

### CODE REVIEW

**Comment by abarth** [comment]
@Hixie Initial take.  Still needs tests (and lots of docs).

File: packages/flutter/lib/src/rendering/sliver_block.dart [hunk 1]
```diff
@@ -1,307 +1,22 @@
...
-    assert(child.hasSize);
-    switch (constraints.axis) {
-      case Axis.horizontal:
-        return child.size.width;
-      case Axis.vertical:
-        return child.size.height;
-    }
-    return null;
-  }
+class RenderSliverBlock extends RenderSliverMultiBoxAdaptor {
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/sliver_block.dart]: maybe we should rename this so it doesn't use the name Block while we're at it.

File: packages/flutter/lib/src/rendering/sliver_list.dart [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+      final int leadingGarbage = math.max(0, firstIndex - oldFirstIndex);
+      final int trailingGarbage = math.max(0, oldLastIndex - targetLastIndex);
+      if (leadingGarbage + trailingGarbage > 0)
+        collectGarbage(leadingGarbage, trailingGarbage);
+    }
+
+    if (firstChild == null) {
+      if (!addInitialChild(index: firstIndex, scrollOffset: _indexToScrollOffset(firstIndex))) {
+        // There are no children.
+        geometry = new SliverGeometry(
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/sliver_list.dart]: i think these are the default values for those arguments, you can just omit them

File: packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+  /// specified in the [constraints]. For example if the children are the
+  /// alphabet, then if [constraints.growthDirection] is
+  /// [GrowthDirection.forward] then index zero is A, and index 25 is Z. On the
+  /// other hand if [constraints.growthDirection] is [GrowthDirection.reverse]
+  /// then index zero is Z, and index 25 is A.
+  ///
+  /// During a call to [createChild] it is valid to remove other children from
+  /// the [RenderSliverBlock] object if they were not created during this frame
+  /// and have not yet been updated during this frame. It is not valid to add
+  /// any children to this render object.
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: except the one being created, presumably

File: packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+  /// for indices adjacent to previously requested indices. It is possible for
+  /// negative indices to be requested. For example: if the user scrolls from
+  /// child 0 to child 10, and then those children get much smaller, and then
+  /// the user scrolls back up again, this method will eventually be asked to
+  /// produce a child for index -1.
+  ///
+  /// If no child corresponds to `index`, then do nothing.
+  ///
+  /// Which child is indicated by index zero depends on the [GrowthDirection]
+  /// specified in the [constraints]. For example if the children are the
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: these hyperlinks are broken now. you need to fully qualify them.

File: packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+  ///
+  /// During a call to [createChild] it is valid to remove other children from
+  /// the [RenderSliverBlock] object if they were not created during this frame
+  /// and have not yet been updated during this frame. It is not valid to add
+  /// any children to this render object.
+  void createChild(int index, { @required RenderBox after });
+
+  /// Remove the given child from the child list.
+  ///
+  /// Called by [collectGarbage], which itself is called from [performLayout],
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: broken links

File: packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+  /// after the layout algorithm has finished and the non-visible children are
+  /// to be removed.
+  ///
+  /// The default implementation calls [remove], which removes the child from
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: there's no default implementation anymore, so this paragraph is wrong

File: packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+  /// the child list.
+  ///
+  /// The index of the given child can be obtained using the [indexOf] method,
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: broken link

File: packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+// ///   been laid out during that layout pass.
+// /// - Children cannot be added except during a call to [allowAdditionsFor], and
+// ///   then only if there is no child correspending to that index (or the child
+// ///   child corresponding to that index was first removed).
+// ///
+// /// ## Writing a RenderSliverBlock subclass
+// ///
+// /// There are three methods to override:
+// ///
+// /// - [createChild], which must create the child for the given index, and then
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: this needs updating

File: packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+      child.parentData = new SliverMultiBoxAdaptorParentData();
+  }
+
+  @protected
+  RenderSliverBoxChildManager get childManager => _childManager;
+  final RenderSliverBoxChildManager _childManager;
+
+  int _currentlyUpdatingChildIndex;
+
+  void allowAdditionsFor(int index, VoidCallback callback) {
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: shouldn't this be protected too

File: packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart [hunk 10]
```diff
@@ -0,0 +1,407 @@
...
+        child = childAfter(child);
+      }
+      return true;
+    });
+  }
+
+  /// Called during layout to create and add the child with the given index and scroll
+  /// offset.
+  ///
+  /// Calls [createChild] to actually create and add the child.
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: broken links

File: packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart [hunk 11]
```diff
@@ -0,0 +1,407 @@
...
+  ///
+  /// Returns false if createChild did not add any child, otherwise returns
+  /// true.
+  ///
+  /// Does not layout the new child.
+  ///
+  /// When this is called, there are no children, so no children can be removed
+  /// during the call to createChild. No child should be added during that call
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: [createChild] (which will then be broken...)

File: packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart [hunk 12]
```diff
@@ -0,0 +1,407 @@
...
+          result = true;
+        } else {
+          result = false;
+        }
+      });
+    });
+    return result;
+  }
+
+  /// Called during layout to create, add, and layout the child before
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: update docs here too

File: packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart [hunk 13]
```diff
@@ -0,0 +1,407 @@
...
+      assert(parentUsesSize);
+      firstChild.layout(childConstraints, parentUsesSize: parentUsesSize);
+      SliverMultiBoxAdaptorParentData firstChildParentData = firstChild.parentData;
+      firstChildParentData.scrollOffset = endScrollOffset - paintExtentOf(firstChild);
+      return firstChild;
+    }
+    return null;
+  }
+
+  /// Called during layout to create, add, and layout the child after
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: and here

File: packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart [hunk 14]
```diff
@@ -0,0 +1,407 @@
...
+    assert(child != null);
+    final SliverMultiBoxAdaptorParentData childParentData = child.parentData;
+    assert(childParentData.index != null);
+    return childParentData.index;
+  }
+
+  /// Returns the scroll offset of the given child, as given by the
+  /// [SliverMultiBoxAdaptorParentData.scrollOffset] field of the child's [parentData].
+  @protected
+  double offsetOf(RenderBox child) {
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: offsetOf and indexOf should both either be protected or not; right now it's inconsistent

File: packages/flutter/lib/src/widgets/sliver.dart [hunk 15]
```diff
@@ -149,33 +96,55 @@ class SliverBlock extends RenderObjectWidget {
   }
 }
 
-class _SliverBlockElement extends RenderObjectElement {
-  _SliverBlockElement(SliverBlock widget) : super(widget);
+class SliverBlock extends SliverMultiBoxAdaptorWidget {
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/sliver.dart]: we should rename this Block too

File: packages/flutter/lib/src/widgets/sliver.dart [hunk 16]
```diff
@@ -149,33 +96,55 @@ class SliverBlock extends RenderObjectWidget {
...
   @override
-  void unmount() {
-    super.unmount();
-    renderObject._element = null;
+  void updateRenderObject(BuildContext context, RenderSliverList renderObject) {
+    renderObject.itemExtent = itemExtent;
   }
+}
+
+class SliverMultiBoxAdaptorElement extends RenderObjectElement implements RenderSliverBoxChildManager{
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/sliver.dart]: nit: space before {

File: packages/flutter/lib/src/widgets/sliver.dart [hunk 17]
```diff
@@ -203,7 +172,7 @@ class _SliverBlockElement extends RenderObjectElement {
       // block widget.)
       for (int index in _childElements.keys.toList()) {
         Element newChild;
-        renderObject._rebuild(index, () {
+        renderObject.allowAdditionsFor(index, () {
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/sliver.dart]: i would feel more comfortable if we buried this call in the SliverMultiBoxAdaptorWidget class somehow

**Comment by Hixie** [comment]
LGTM
modulo the tests and docs you said you wanted to add :-)

File: packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart (hunk 9 above)
Comment (abarth) [inline packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: I call this from the element.  Before there was a private trampoline to call it.

File: packages/flutter/lib/src/rendering/sliver_block.dart (hunk 1 above)
Comment (abarth) [inline packages/flutter/lib/src/rendering/sliver_block.dart]: I'll do the rename in a separate patch to keep the history a bit cleaner.

File: packages/flutter/lib/src/rendering/sliver_list.dart (hunk 2 above)
Comment (abarth) [inline packages/flutter/lib/src/rendering/sliver_list.dart]: Done.

File: packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart (hunk 14 above)
Comment (abarth) [inline packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: Done.

File: packages/flutter/lib/src/widgets/sliver.dart (hunk 16 above)
Comment (abarth) [inline packages/flutter/lib/src/widgets/sliver.dart]: Done.

**Comment by abarth** [comment]
Wrote tests, found bugs.

File: packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart (hunk 3 above)
Comment (abarth) [inline packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: Done.

File: packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart (hunk 4 above)
Comment (abarth) [inline packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: Done.

File: packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart (hunk 5 above)
Comment (abarth) [inline packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: done.

File: packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart (hunk 6 above)
Comment (abarth) [inline packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: Done.

File: packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart (hunk 7 above)
Comment (abarth) [inline packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: Done.

File: packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart (hunk 8 above)
Comment (abarth) [inline packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: Done.

File: packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart (hunk 10 above)
Comment (abarth) [inline packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: Done.

File: packages/flutter/lib/src/widgets/sliver.dart (hunk 17 above)
Comment (abarth) [inline packages/flutter/lib/src/widgets/sliver.dart]: I buried this function entirely in SliverMultiBoxAdaptorElement.

**Comment by Hixie** [comment]
This technically sent us over our aot_snapshot_size_rodata budget, but I'm going to say that's ok because we're going to be deleting a lot of code soon because of this patch.

**Pull request merged**